src/main/java/at/ac/tuwien/kr/alpha/solver/TrailAssignment.java

```diff
@@ -0,0 +1,727 @@
+	 */
+	public int getOutOfOrderDecisionLevel(int atom) {
+		int lowestDecisionLevel = Integer.MAX_VALUE;
+		for (OutOfOrderLiteral outOfOrderLiteral : outOfOrderLiterals) {
+			if (outOfOrderLiteral.atom == atom && outOfOrderLiteral.decisionLevel < lowestDecisionLevel) {
+				lowestDecisionLevel = outOfOrderLiteral.decisionLevel;
+			}
+		}
+		return lowestDecisionLevel;
+	}
+
+	public int getOutOfOrderStrongDecisionLevel(int atom) {
+		int lowestDecisionLevel = Integer.MAX_VALUE;
+		for (OutOfOrderLiteral outOfOrderLiteral : outOfOrderLiterals) {
+			if (outOfOrderLiteral.atom == atom && outOfOrderLiteral.value == TRUE && outOfOrderLiteral.decisionLevel < lowestDecisionLevel) {
+				lowestDecisionLevel = outOfOrderLiteral.decisionLevel;
+			}
+		}
+		return lowestDecisionLevel;
+	}
+
+	private void informCallback(Integer atom) {
+		if (callbackUponChange[atom]) {
+			choiceManagerCallback.callbackOnChanged(atom);
+		}
+	}
+
+	private void removeLastDecisionLevel() {
+		// Remove all atoms recorded in the highest decision level.
+		Integer start = trailIndicesOfDecisionLevels.get(getDecisionLevel());
+		ListIterator<Integer> backtrackIterator = trail.listIterator(start);
+		while (backtrackIterator.hasNext()) {
+			Integer backtrackAtom = backtrackIterator.next();
+			// Skip already backtracked atoms.
+			if (getTruth(backtrackAtom) == null) {
+				continue;
+			}
+			if (getWeakDecisionLevel(backtrackAtom) < getDecisionLevel()) {
+				// Restore TRUE to MBT if this was assigned at a lower level.
+				if (getTruth(backtrackAtom) != TRUE) {
+					throw oops("Backtracking assignment with lower decision level whose value is not TRUE.");
+				}
+				values[backtrackAtom] = (getWeakDecisionLevel(backtrackAtom) << 2) | translateTruth(MBT);
+				mbtCount++;
+			} else {
+				if (getTruth(backtrackAtom) == MBT) {
+					mbtCount--;
+				}
+				values[backtrackAtom] = 0;
+			}
+			strongDecisionLevels[backtrackAtom] = -1;
+			informCallback(backtrackAtom);
+		}
+		// Remove atoms from trail.
+		while (trail.size() > start) {
+			trail.remove(trail.size() - 1);
+		}
+		trailIndicesOfDecisionLevels.remove(trailIndicesOfDecisionLevels.size() - 1);
+
+		int trailCurrentIndex = trail.size();
+		int trailPrevIndex = trailIndicesOfDecisionLevels.size() < 1  ? 0 : trailIndicesOfDecisionLevels.get(trailIndicesOfDecisionLevels.size() - 1);
+		currentPropagationLevel = trailCurrentIndex - trailPrevIndex; // Note: due to MBT->TRUE assignments this may be higher than the actual decision level, but never below.
+	}
+
+
+	int replayCounter;
+	private void replayOutOfOrderLiterals() {
+		// Replay out-of-order assigned literals.
+		if (highestDecisionLevelContainingOutOfOrderLiterals >= getDecisionLevel()) {
+			// Replay all still-valid literals and remove invalid ones.
+			LOGGER.trace("Replaying out-of-order literals.");
+			int k = 0; // counter for out-of-order literals to keep further.
+			for (int i = 0; i < outOfOrderLiterals.size(); i++) {
```

Comment:
Enhanced for loop could be used instead: `for (OutOfOrderLiteral outOfOrderLiteral : outOfOrderLiterals)`